flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerCleanupITCase.java

```diff
@@ -0,0 +1,298 @@
+							source.setInvokableClass(FailingBlockingInvokable.class);
+						} else {
+							source.setInvokableClass(NoOpInvokable.class);
+						}
+						source.setParallelism(num_tasks);
+
+						JobGraph jobGraph = new JobGraph("BlobCleanupTest", source);
+						final JobID jid = jobGraph.getJobID();
+
+						// request the blob port from the job manager
+						Future<Object> future = jobManagerGateway
+							.ask(JobManagerMessages.getRequestBlobManagerPort(), remaining());
+						int blobPort = (Integer) Await.result(future, remaining());
+
+						// upload a blob
+						BlobKey key1;
+						bc = new BlobClient(new InetSocketAddress("localhost", blobPort),
+							config);
+						try {
+							key1 = bc.put(jid, new byte[10]);
+						} finally {
+							bc.close();
+						}
+						jobGraph.addBlob(key1);
+
+						if (testCase == TestCase.JOB_SUBMISSION_FAILS) {
+							// add an invalid key so that the submission fails
+							jobGraph.addBlob(new BlobKey());
+						}
+
+						// Submit the job and wait for all vertices to be running
+						jobManagerGateway.tell(
+							new JobManagerMessages.SubmitJob(
+								jobGraph,
+								ListeningBehaviour.EXECUTION_RESULT),
+							testActorGateway);
+						if (testCase == TestCase.JOB_SUBMISSION_FAILS) {
+							expectMsgClass(JobManagerMessages.JobResultFailure.class);
+						} else {
+							expectMsgClass(JobManagerMessages.JobSubmitSuccess.class);
+
+							if (testCase == TestCase.JOB_FAILS) {
+								// fail a task so that the job is going to be recovered (we actually do not
+								// need the blocking part of the invokable and can start throwing right away)
+								FailingBlockingInvokable.unblock();
+
+								// job will get restarted, BlobCache may re-download the BLOB if already deleted
+								// then the tasks will fail again and the restart strategy will finalise the job
+
+								expectMsgClass(JobManagerMessages.JobResultFailure.class);
+							} else if (testCase == TestCase.JOB_IS_CANCELLED) {
+								jobManagerGateway.tell(
+									new JobManagerMessages.CancelJob(jid),
+									testActorGateway);
+								expectMsgClass(JobManagerMessages.CancellationResponse.class);
+
+								// job will be cancelled and everything should be cleaned up
+
+								expectMsgClass(JobManagerMessages.JobResultFailure.class);
+							} else {
+								expectMsgClass(JobManagerMessages.JobResultSuccess.class);
+							}
+						}
+
+						// both BlobServer and BlobCache should eventually delete all files
+
+						File[] blobDirs = blobBaseDir.listFiles(new FilenameFilter() {
+							@Override
+							public boolean accept(File dir, String name) {
+								return name.startsWith("blobStore-");
+							}
+						});
+						assertNotNull(blobDirs);
+						for (File blobDir : blobDirs) {
+							waitForEmptyBlobDir(blobDir, remaining());
+						}
+
+					} catch (Exception e) {
+						e.printStackTrace();
+						fail(e.getMessage());
```

Comment:
Unfortunately, we cannot throw an arbitrary exception from `JavaTestKit.Within#run()`, otherwise it would be nice.